File: Graph/leetcode_684.py
# Problem Link == == >>https://leetcode.com/problems/redundant-connection/description/

class Solution:
    def findRedundantConnection(self, edges: list[list[int]]) -> list[int]:
        n = len(edges)
        graph = [[] for _ in range(n + 1)]  # Adjacency list representation
        visited = [False] * (n + 1)
        cycle = set()
        cycle_start = -1

        # Build the graph from the edges
        for edge in edges:
            graph[edge[0]].append(edge[1])
            graph[edge[1]].append(edge[0])

        def dfs(cur, par=-1):
            nonlocal cycle_start
            if visited[cur]:
                cycle_start = cur
                return  # Found a cycle by reaching a visited node

            visited[cur] = True
            for child in graph[cur]:
                if child == par:
                    continue  # Skip parent to avoid back-and-forth loop
                if not cycle:
                    dfs(child, cur)
                if cycle_start != -1:
                    # Add current node to cycle set if cycle detected
                    cycle.add(cur)
                if cur == cycle_start:
                    cycle_start = -1
                    return  # Reached the start of the cycle, stop exploration

        # Perform DFS to detect cycles and populate the cycle set
        dfs(1)

        # Find the last edge in the input that has both nodes in the cycle
        for i in range(n - 1, -1, -1):
            if edges[i][0] in cycle and edges[i][1] in cycle:
                return edges[i]

        return []  # No redundant connection found


# Main function for testing
edges = [[1, 2], [1, 3], [2, 3]]
solution = Solution()
result = solution.findRedundantConnection(edges)

if not result:
    print("No redundant connection found.")
else:
    print(f"Redundant connection: {result[0]}, {result[1]}")

File: Graph/leetcode_684.cpp

// Problem Link == == >>https://leetcode.com/problems/redundant-connection/description/

#include<bits/stdc++.h>
using namespace std;

class Solution
{
public:
    vector<int> findRedundantConnection(vector<vector<int>> &edges)
    {
        int n = edges.size();
        vector<vector<int>> graph(n + 1);
        vector<bool> visited(n + 1, false);
        unordered_set<int> cycle;
        int cycleStart = -1;

        // Build the graph from the edges
        for (auto &edge : edges)
        {
            graph[edge[0]].push_back(edge[1]);
            graph[edge[1]].push_back(edge[0]);
        }

        // Perform DFS to detect cycles and populate the cycle set
        dfs(graph, visited, cycle, cycleStart, 1);

        // Find the last edge in the input that has both nodes in the cycle
        for (int i = n - 1; i >= 0; --i)
        {
            if (cycle.count(edges[i][0]) && cycle.count(edges[i][1]))
            {
                return edges[i];
            }
        }

        return {}; // No redundant connection found
    }

private:
    void dfs(vector<vector<int>> &graph, vector<bool> &visited, unordered_set<int> &cycle, int &cycleStart, int cur, int par = -1)
    {
        if (visited[cur])
        {
            cycleStart = cur;
            return; // Found a cycle by reaching a visited node
        }

        visited[cur] = true;
        for (auto child : graph[cur])
        {
            if (child == par)
                continue; // Skip parent to avoid back-and-forth loop
            if (cycle.empty())
            {
                dfs(graph, visited, cycle, cycleStart, child, cur);
            }
            if (cycleStart != -1)
            {
                cycle.insert(cur); // Add current node to cycle set if cycle detected
            }
            if (cur == cycleStart)
            {
                cycleStart = -1;
                return; // Reached the start of the cycle, stop exploration
            }
        }
    }
};

int main()
{
    vector<vector<int>> edges = {{1, 2}, {1, 3}, {2, 3}};
    Solution s;
    vector<int> result = s.findRedundantConnection(edges);

    if (result.empty())
    {
        cout << "No redundant connection found." << endl;
    }
    else
    {
        cout << "Redundant connection: [" << result[0] << ", " << result[1] << "]" << endl;
    }

    return 0;
}
